### PR TITLE
Use route rather than use thus reducing the number of stack frames

### PR DESCRIPTION
**Pull request by zeripath**
Since the move to Chi the number of stack frames has proliferated somewhat catastrophically and we're up to 96 frames with multiple tests of the url outside of a trie which is inefficient.

This PR reduces the number of stack frames by 6 through careful use of Route, moves Captcha into its own router so that it only fires on Captcha routes, similarly for avatars and repo-avatars.

The `robots.txt`, `/` and `apple-touch-icon.png` are moved out of requiring Contexter.

It moves access logger higher in the stack frame because there is no reason why it can't be higher.

Extract from #15186
Contains #15292

### CODE REVIEW

File: routers/routes/web.go [hunk 1]
```diff
@@ -141,101 +143,93 @@ func WebRoutes() *web.Route {
...
+
+	// for health check - doeesn't need to be passed through gzip handler
+	routes.Head("/", func(w http.ResponseWriter, req *http.Request) {
+		w.WriteHeader(http.StatusOK)
+	})
+
+	// this png is very likely to always be below the limit for gzip so it doesn't need to pass through gzip
+	routes.Get("/apple-touch-icon.png", func(w http.ResponseWriter, req *http.Request) {
+		http.Redirect(w, req, path.Join(setting.StaticURLPrefix, "img/apple-touch-icon.png"), 301)
+	})
```
Comment (silverwind) [inline routers/routes/web.go:177]: Can probably nuke that route and make it use a `<link>` tag to directly link to the image. I recall this route was only there for some edge case with Firefox, but it should be favoring the SVG icon over this one nowadays.

**Review by KN4CK3R — APPROVED** [review]

**Review by silverwind — APPROVED** [review]

**Comment by zeripath** [comment]
Make lgtm work

**Comment by wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf** [comment]
after upgrading to this the user-specified avatars won't load.. I'm not sure if that's intentional, it's potentially `kind/breaking`.

**Comment by zeripath** [comment]
Not intentional at all - I thought I'd tested all of this. 

It might be that it's reverted your recent fix as this was written before it.

Do you remember your pr that fixed things?

**Comment by wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf** [comment]
> Not intentional at all - I thought I'd tested all of this.

No worries.

> It might be that it's reverted your recent fix as this was written before it.
> 
> Do you remember your pr that fixed things?

There was this thing with assets, should I check that?
I'll go revert that *my fix* (moving things to `/assets/{img,css,*}/...`) and report back.

**Comment by wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf** [comment]
looks like custom css still works fine and so do custom `/assets`-based icons.
this is different, though, this does not pertain customizations, rather `/attachments` stuff

**Comment by zeripath** [comment]
Unfortunately there was quite a few prs between this being written and it getting approved - I bet something to do with the assets move is the cause.

I should have rechecked before merging.

I do think this is really worth doing so fixing the problem rather than reverting this pr should be the aim.

**Comment by wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf** [comment]
I have done a quick write-up of the issue so that I don't pollute this PR anymore

**Pull request merged**